### PR TITLE
Set iOS default frame rate to screen max.

### PR DESCRIPTION
**Pull request by cyanglaz**
The default iOS frame rate is set to 60 by default. This PR explicitly sets the preferred rate to the maximum refresh rate supported by the display. (120 on iPhone 13 pro, 60 on slower devices)
On iOS 15 and above, we also set the frame rate range, where the preferred and max are both screen max, the min is max/2.
Also, we avoid setting the preferred frame rate to a value lower than 60.

Fixes https://github.com/flutter/flutter/issues/90675

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by chinmaygarde — DISMISSED** [review]
During the API design review presented by @iskakaushik, this didn't seem like the direction we were going to take.

**Comment by cyanglaz** [comment]
@chinmaygarde After the meeting, we had an offline discussion and we wanted to enable 120 first as an intermediate step. This will be a beta feature which user has to opt-in.

**Comment by chinmaygarde** [comment]
I am not convinced that is a good idea. During the review, there was widespread opposition to this approach. And doing just that but only for temporarily and for users that opt into it via a flag isn't something we would want to do. We have also had confusion in the past about when certain flags would be necessary (like those for platform views). Can you fill me in on the context of the discussion that led to making this decision?

**Comment by cyanglaz** [comment]
The flag would be necessary for iOS to enable 120HZ nonetheless. So user has to opt-in. 

The reasons we wanted to do this first: 
1. We'd like to resolve https://github.com/flutter/flutter/issues/90675 soon enough so apps who want 120 hz can opt-in (via Apple's required info.plist)
2. We are still deciding on the exact logics on the engine's heuristics. (Currently we have: https://github.com/flutter/engine/pull/29692)

Also see Kaushik's comment in https://github.com/flutter/flutter/issues/90675#issuecomment-971863161

**Comment by chinmaygarde** [comment]
> necessary for iOS to enable 120HZ nonetheless. So user has to opt-in.

Not sure I understand why if progress is made on (2) right now. This patch as a solution for (1) seems like an odd choice given the degree of opposition to this approach during the review.

**Comment by chinmaygarde** [comment]
Just chatted with Kaushik about this and I was confused about the flag referred to here. To clarify, it was `CADisableMinimumFrameDurationOnPhone` and not one of our own (which is what I thought it was). It also doesn't seem to be always capping to maximum. Sorry about the confusion.

**Comment by cyanglaz** [comment]
CI blocked by https://github.com/flutter/flutter/issues/85555

**Comment by Hixie** [comment]
here's a proposed api for the framework: https://github.com/flutter/flutter/compare/master...Hixie:freq

**Comment by tilemoon** [comment]
any progress now?

**Comment by Void48** [comment]
Please ☹️

**Comment by pkoehler-altow-extern** [comment]
Push :)

**Comment by leonidlist** [comment]
Any updates on that?

**Comment by luckysmg** [comment]
Is this feature is developing now?

**Comment by cyanglaz** [comment]
Hi Everyone. Thank you for being patient. This is in development and we are waiting on some [benchmark work](https://github.com/flutter/flutter/pull/95699) to be done before landing this. 

I will hide the comments that unrelated to the solution to keep the conversation in this PR directly related to the solution itself. Thank you for your understanding :)

**Comment by luckysmg** [comment]
I wrote some code to try ,both in flutter framework and engine ,any advices😀?
https://github.com/flutter/flutter/pull/96710
https://github.com/flutter/engine/pull/30900

**Comment by cyanglaz** [comment]
Since it is still blocked by https://github.com/flutter/flutter/issues/85555
I'm going to temporarily close this issue to avoid unnecessary surfacing in triage.

**Comment by jmagman** [comment]
iOS 15 SDK is now available in the engine https://github.com/flutter/flutter/issues/85555

**Comment by anuccio15** [comment]
@jmagman awesome work. Any idea when we can expect to have the refresh rate support in flutter for iOS merged?

**Comment by jmagman** [comment]
@iskakaushik and @cyanglaz are unblocked to continue working on this, we don't have an ETA yet.

**Comment by cyanglaz** [comment]
Reopen as the SDK is available for the code in this PR to compile. https://github.com/flutter/flutter/issues/90675#issuecomment-1082421062

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm:82]: oops, good catch! ty, will fix

Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm]: Fixed, PTAL

**Comment by cyanglaz** [comment]
@iskakaushik Added tests, please take a look

File: shell/platform/darwin/ios/BUILD.gn
Comment (jmagman) [inline shell/platform/darwin/ios/BUILD.gn:211]: Why does this test need to use mrc? Can it be moved to `ios_test_flutter`?  Your test as written isn't releasing everything.

File: shell/platform/darwin/ios/framework/Source/VsyncWaiterIosTest.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/VsyncWaiterIosTest.mm:32]: I don't think this test is very valuable since your other tests validate that it's not nil. Can you instead test the case where `CADisableMinimumFrameDurationOnPhone` is not present into the plist?

File: shell/platform/darwin/ios/BUILD.gn
Comment (cyanglaz) [reply shell/platform/darwin/ios/BUILD.gn:211]: I got compiling issue trying to use ARC due to `scoped_nsobject` used in the header. `"flutter/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.h"`

File: shell/platform/darwin/ios/framework/Source/VsyncWaiterIosTest.mm
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/VsyncWaiterIosTest.mm:32]: done

File: shell/platform/darwin/ios/framework/Source/VsyncWaiterIosTest.mm
Comment (iskakaushik) [inline shell/platform/darwin/ios/framework/Source/VsyncWaiterIosTest.mm:27]: typo:`Variable`

**Review by iskakaushik — APPROVED** [review]
LGTM

**Comment by cyanglaz** [comment]
@chinmaygarde your old review is still blocking the PR. I think @iskakaushik answered your question offline. Do you mind take another look?

**Comment by zanderso** [comment]
Chinmay is offline today. Based on his comment about the offline discussions, I think he is on board, and I dismissed his review so that this can land.

**Review by jmagman — APPROVED** [review]

**Pull request merged**
